tests/functional/history/test_db.py

```diff
@@ -0,0 +1,199 @@
+import tempfile
+import shutil
+import json
+import os
+
+from awscli.customizations.history.db import DatabaseConnection
+from awscli.customizations.history.db import DatabaseRecordWriter
+from awscli.customizations.history.db import DatabaseRecordReader
+from awscli.testutils import unittest
+from awscli.compat import sqlite3
+
+
+class BaseDatabaseTest(unittest.TestCase):
+    def setUp(self):
+        self.tempdir = tempfile.mkdtemp()
+        self.db_filename = os.path.join(self.tempdir, 'history.db')
+        self.connection = DatabaseConnection(db_filename=self.db_filename)
+        self.connection.row_factory = sqlite3.Row
+
+    def tearDown(self):
+        shutil.rmtree(self.tempdir)
+
+    def _get_cursor(self):
+        return self.connection.cursor()
+
+
+@unittest.skipIf(sqlite3 is None,
+                 "sqlite3 not supported in this python")
+class TestDatabaseRecordWriter(BaseDatabaseTest):
+    def test_does_create_database(self):
+        self.assertTrue(os.path.isfile(self.db_filename))
+
+    def test_does_create_table(self):
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM sqlite_master WHERE "
+                       "type='table' AND name='records';")
+        result = cursor.fetchone()
+        self.assertEqual(result[0], 1)
+
+    def test_can_write_record(self):
+        writer = DatabaseRecordWriter(connection=self.connection)
+        known_record_fields = {
+            'source': 'TEST',
+            'event_type': 'foo',
+            'payload': {"foo": "bar"},
+            'timestamp': 1234
+        }
+        writer.write_record(known_record_fields)
+
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM records;")
+        num_records = cursor.fetchone()
+        self.assertEqual(num_records[0], 1)
+
+        cursor.execute("SELECT * FROM records;")
+        record = dict(cursor.fetchone())
+        for col_name, row_value in known_record_fields.items():
+            # Normally our reader would take care of parsing the JSON from
+            # the payload.
+            if col_name == 'payload':
+                record[col_name] = json.loads(record[col_name])
+            self.assertEqual(record[col_name], row_value)
+
+        self.assertIn('id', record.keys())
+        self.assertIn('request_id', record.keys())
+
+    def test_can_write_many_records(self):
+        writer = DatabaseRecordWriter(connection=self.connection)
+        known_record_fields = {
+            'source': 'TEST',
+            'event_type': 'foo',
+            'payload': '',
+            'timestamp': 12334
+        }
+        records_to_write = 40
+        for _ in range(records_to_write):
+            writer.write_record(known_record_fields)
+
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM records;")
+        num_records = cursor.fetchone()
+        self.assertEqual(num_records[0], records_to_write)
+
```

Comment:
I wonder if it makes sense to put some thread property based testing here in the functional tests. The idea is you would use a single db connection behind a lock and then have threads write to the db and at the end you would take the reader to make sure that all of the records were there and there was X distrinct request id's where X would be the number of http lifecycles multiplied by number of threads. Furthermore, you could do some more testing to make sure that when if you group them together by request id's all of the records are as expected.

Reply:
Yea that sounds good to me. I only have the one thread related test so there could definitely be a lot more.